### PR TITLE
refactor: move member adding to manager

### PR DESCRIPTION
**Pull request by ckohen**
**Please describe the changes this PR makes and why it should be merged:**

The long awaited move is here.
Also comes with a bug fix to the method (throws properly when a role is invalid), along with no more param mutation.
Further adds a force param for those of you that don't get member events (if a member has left while the bot is running, they still are cached 😢 )

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: src/managers/GuildMemberManager.js [hunk 1]
```diff
@@ -66,6 +66,49 @@ class GuildMemberManager extends CachedManager {
...
+   * Adds a user to the guild using OAuth2. Requires the `CREATE_INSTANT_INVITE` permission.
+   * @param {UserResolvable} user The user to add to the guild
+   * @param {AddGuildMemberOptions} options Options for adding the user to the guild
+   * @returns {Promise<GuildMember>}
+   */
+  async add(user, options) {
+    const userId = this.client.users.resolveId(user);
+    if (!userId) throw new TypeError('INVALID_TYPE', 'user', 'UserResolvable');
+    if (!options?.force) {
+      if (this.cache.has(userId)) return this.cache.get(userId);
```
Comment (kyranet) [inline src/managers/GuildMemberManager.js:92]: Avoid has-set:
```suggestion
      const cachedUser = this.cache.get(userId);
      if (cachedUser) return cachedUser;
```

File: src/managers/GuildMemberManager.js [hunk 2]
```diff
@@ -66,6 +66,49 @@ class GuildMemberManager extends CachedManager {
...
+      for (const role of options.roles.values()) {
+        const resolvedRole = this.guild.roles.resolveId(role);
+        if (!resolvedRole) throw new TypeError('INVALID_ELEMENT', 'Array or Collection', 'options.roles', role);
+        resolvedRoles.push(resolvedRole);
+      }
+      options.roles = resolvedRoles;
+    }
+    const data = await this.client.api.guilds(this.guild.id).members(userId).put({ data: options });
+    // Data is an empty buffer if the member is already part of the guild.
+    return data instanceof Buffer ? this.members.fetch(userId) : this._add(data);
```
Comment (kyranet) [inline src/managers/GuildMemberManager.js:109]: I know this is the same behaviour as before, but we should probably make the auto-fetch opt-in in the future.

**Review by kyranet — CHANGES_REQUESTED** [review]
Just one change:

File: src/managers/GuildMemberManager.js [hunk 3]
```diff
@@ -66,6 +66,49 @@ class GuildMemberManager extends CachedManager {
...
+  /**
+   * Adds a user to the guild using OAuth2. Requires the `CREATE_INSTANT_INVITE` permission.
+   * @param {UserResolvable} user The user to add to the guild
+   * @param {AddGuildMemberOptions} options Options for adding the user to the guild
+   * @returns {Promise<GuildMember>}
+   */
+  async add(user, options) {
+    const userId = this.client.users.resolveId(user);
+    if (!userId) throw new TypeError('INVALID_TYPE', 'user', 'UserResolvable');
+    if (!options?.force) {
```
Comment (NotSugden) [inline src/managers/GuildMemberManager.js:91]:
```suggestion
    if (!options.force) {
```
options is documented as required

File: src/managers/GuildMemberManager.js (hunk 2 above)
Comment (vladfrangu) [reply src/managers/GuildMemberManager.js:109]:
```suggestion
    return data instanceof Buffer ? this.fetch(userId) : this._add(data);
```

**Review by vladfrangu — DISMISSED** [review]

File: src/managers/GuildMemberManager.js [hunk 4]
```diff
@@ -66,6 +66,55 @@ class GuildMemberManager extends CachedManager {
...
+        const resolvedRole = this.guild.roles.resolveId(role);
+        if (!resolvedRole) throw new TypeError('INVALID_ELEMENT', 'Array or Collection', 'options.roles', role);
+        resolvedRoles.push(resolvedRole);
+      }
+      options.roles = resolvedRoles;
+    }
+    const data = await this.client.api.guilds(this.guild.id).members(userId).put({ data: options });
+    // Data is an empty buffer if the member is already part of the guild.
+    let newUser = data instanceof Buffer ? null : this._add(data);
+    if (!newUser && (options.fetchWhenExisting ?? true)) {
```
Comment (NotSugden) [inline src/managers/GuildMemberManager.js:112]:
```suggestion
    if (!newUser && options.fetchWhenExisting !== false) {
```
i'd personally do this but either works really

**Review by NotSugden — COMMENTED** [review]
its probably not important, but `fetchWhenExisting` and `accessToken` are being sent to the API when neither fields are valid (i don't believe it causes an error), just thought it worth noting

File: src/managers/GuildMemberManager.js [hunk 5]
```diff
@@ -66,6 +66,60 @@ class GuildMemberManager extends CachedManager {
...
+    if (!options.force) {
+      const cachedUser = this.cache.get(userId);
+      if (cachedUser) return cachedUser;
+    }
+    const resolvedOptions = {
+      nick: options.nick,
+      mute: options.mute,
+      deaf: options.deaf,
+    };
+    resolvedOptions.access_token = options.accessToken;
```
Comment (kyranet) [inline src/managers/GuildMemberManager.js:101]:
```suggestion
      deaf: options.deaf,
      access_token: options.accessToken,
    };
```

File: src/managers/GuildMemberManager.js [hunk 6]
```diff
@@ -66,6 +66,60 @@ class GuildMemberManager extends CachedManager {
...
+      }
+      resolvedOptions.roles = resolvedRoles;
+    }
+    const data = await this.client.api.guilds(this.guild.id).members(userId).put({ data: resolvedOptions });
+    // Data is an empty buffer if the member is already part of the guild.
+    let newUser = data instanceof Buffer ? null : this._add(data);
+    if (!newUser && options.fetchWhenExisting !== false) {
+      newUser = this.fetch(userId);
+    }
+    return newUser;
```
Comment (kyranet) [inline src/managers/GuildMemberManager.js:120]: You could probably do something like this, I don't think `_add` returns a falsy value.
```suggestion
    return data instanceof Buffer
    	? options.fetchWhenExisting === false
    		? null
    		: this.fetch(userId)
    	: this._add(data);
```

Prettier will most likely complain about formatting, so please do this in VSCode 🙏🏼

File: typings/index.d.ts [hunk 7]
```diff
@@ -2333,6 +2332,8 @@ export class GuildManager extends CachedManager<Snowflake, Guild, GuildResolvabl
 export class GuildMemberManager extends CachedManager<Snowflake, GuildMember, GuildMemberResolvable> {
   public constructor(guild: Guild, iterable?: Iterable<unknown>);
   public guild: Guild;
+  public add(user: UserResolvable, options: AddGuildMemberOptions & { fetchWhenExisting: true }): Promise<GuildMember>;
+  public add(user: UserResolvable, options: AddGuildMemberOptions): Promise<GuildMember | null>;
```
Comment (kyranet) [inline typings/index.d.ts:2336]: Probably needs to be done this way? `fetchWhenExisting` defaults to `true`, so we need to check for a required `false` value, then the rest of overloads.
```suggestion
  public add(user: UserResolvable, options: AddGuildMemberOptions & { fetchWhenExisting: false }): Promise<GuildMember | null>;
  public add(user: UserResolvable, options: AddGuildMemberOptions): Promise<GuildMember>;
```

You can also add a type test.

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by iCrawl — APPROVED** [review]

File: typings/index.ts [hunk 8]
```diff
@@ -398,13 +398,38 @@ client.on('ready', async () => {
...
   });
+
+  // @ts-expect-error
+  assertIsMember(g.members.add(testUserId));
+
+  // @ts-expect-error
+  assertIsMember(g.members.add(testUserId, {}));
+
+  // @ts-expect-error
+  assertIsMember(g.members.add(testUserId, { accessToken: 'totallyRealAccessToken', roles: [g.roles.cache] }));
```
Comment (kyranet) [inline typings/index.ts:419]: What error does this emit? `fetchWhenExisting` defaults to `true`, so this should not yield an error.

File: typings/index.ts [hunk 9]
```diff
@@ -398,13 +398,38 @@ client.on('ready', async () => {
...
+  // @ts-expect-error
+  assertIsMember(g.members.add(testUserId, { accessToken: 'totallyRealAccessToken', fetchWhenExisting: false }));
+
+  assertIsMember(
+    g.members.add(testUserId, {
+      accessToken: 'totallyRealAccessToken',
+      mute: true,
+      deaf: false,
+      roles: [g.roles.cache.first()!],
+      force: true,
```
Comment (kyranet) [inline typings/index.ts:430]: Probably want to test when it's set to `true`, so we test all three cases (`undefined`, `false`, and `true`):
```suggestion
      roles: [g.roles.cache.first()!],
      fetchWhenExisting: true,
      force: true,
```

File: typings/index.ts [hunk 10]
```diff
@@ -398,13 +398,38 @@ client.on('ready', async () => {
   });
 });
 
+// This is to check that stuff is the right type
+declare const assertIsMember: (m: Promise<GuildMember>) => void;
```
Comment (kyranet) [inline typings/index.ts:402]: Perhaps name this `assertIsPromiseMember`? Although the one below (`assertIsMessage`) should probably be updated.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: typings/index.ts (hunk 8 above)
Comment (ckohen) [reply typings/index.ts:419]: This is an invalid `RoleResolvable`

File: typings/index.ts (hunk 10 above)
Comment (ckohen) [reply typings/index.ts:402]: Do we need to differentiate? I could union type `Promise<GuildMember> | GuildMember`

Comment (kyranet) [reply typings/index.ts]: We want type assertions to be as strict as possible, we probably don't want a method that should be returning a `Promise<T>` to accidentally return `T` 😄

File: typings/index.ts (hunk 8 above)
Comment (kyranet) [reply typings/index.ts:419]: You can write a description next to `// @ts-expect-error`, please do so.

**Review by kyranet — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**